contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,866 @@
+        require(
+            IStakePool(valInfo.poolModule).getPooledBNBByShares(shares) >= minDelegationBNBChange,
+            "INVALID_UNDELEGATION_AMOUNT"
+        );
+
+        uint256 bnbAmount = IStakePool(valInfo.poolModule).undelegate(delegator, shares);
+        emit Undelegated(operatorAddress, delegator, shares, bnbAmount);
+
+        if (
+            delegator == operatorAddress && IStakePool(valInfo.poolModule).getSelfDelegationBNB() < minSelfDelegationBNB
+        ) {
+            _validators[operatorAddress].jailed = true;
+            _removeEligibleValidator(valInfo.consensusAddress);
+            emit ValidatorJailed(operatorAddress);
+        }
+
+        IGovBNB(govBNB).burn(operatorAddress, delegator, bnbAmount);
+    }
+
+    function redelegate(
+        address srcValidator,
+        address dstValidator,
+        uint256 shares
+    )
+        public
+        onlyInitialized
+        validatorExist(srcValidator)
+        validatorExist(dstValidator)
+        validatorNotJailed(dstValidator)
+        whenNotPaused
+    {
+        require(shares > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory srcValInfo = _validators[srcValidator];
+        Validator memory dstValInfo = _validators[dstValidator];
+        require(
+            IStakePool(srcValInfo.poolModule).getPooledBNBByShares(shares) >= minDelegationBNBChange,
+            "INVALID_REDELEGATION_AMOUNT"
+        );
+
+        uint256 bnbAmount = IStakePool(srcValInfo.poolModule).unbond(delegator, shares);
+        uint256 newShares = IStakePool(dstValInfo.poolModule).delegate{ value: bnbAmount }(delegator);
+        emit Redelegated(srcValidator, dstValidator, delegator, shares, newShares, bnbAmount);
+
+        IGovBNB(govBNB).burn(srcValidator, delegator, bnbAmount);
+        IGovBNB(govBNB).mint(dstValidator, delegator, bnbAmount);
+    }
+
+    /**
+     * @dev Claim the undelegated BNB from the pool after unbondPeriod
+     */
+    function claim(
+        address operatorAddress,
+        uint256 requestNumber
+    ) external onlyInitialized validatorExist(operatorAddress) {
+        uint256 bnbAmount = IStakePool(_validators[operatorAddress].poolModule).claim(msg.sender, requestNumber);
+        emit Claimed(operatorAddress, msg.sender, bnbAmount);
+    }
+
+    function upgradePoolImplementation() external onlyInitialized validatorExist(msg.sender) {
+        address operatorAddress = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+        address poolProxy = valInfo.poolModule;
+        require(!_checkPoolImplementation(poolProxy), "NO_NEED_TO_UPGRADE");
+        ProxyAdmin(_proxyAdmin).upgrade(ITransparentUpgradeableProxy(poolProxy), poolImplementation);
+    }
+
+    /*----------------- system functions -----------------*/
+    function distributeReward(address consensusAddress) external payable onlyInitialized onlyValidatorContract {
```

Comment:
`distributeReward` is triggered by consensus engine, we should make it always success, so the `require` here in this function is not allowed.